src/server/replica.cc

```diff
@@ -739,8 +739,11 @@ void Replica::CloseSocket() {
   unique_lock lk(sock_mu_);
   if (sock_) {
     sock_->proactor()->Await([this] {
-      auto ec = sock_->Shutdown(SHUT_RDWR);
-      LOG_IF(ERROR, ec) << "Could not shutdown socket " << ec;
+      if (sock_->IsOpen()) {
```

Comment:
I looked at `LinuxSocketBase::IsOpen`. It checks `fd_ & IS_SHUTDOWN`, and `IS_SHUTDOWN` is `1`. So if `fd_` is `-1` wouldn't the socket still return `true` here?

Reply:
The whole 'stashing state on the file descriptor' looks kinda fishy there, why don't we use bitfields?

Reply:
`-1` makes fd_ all 1's, so `(fd_ & IS_SHUTDOWN)` is 1, which means IsOpen() returns false

Yes, it looks a bit fishy, given that we don't have that much sockets in df we even can waste a byte on proper flags 🤣 I also assume that it won't even matter because of alignment

Reply:
Oops, I just saw `return (fd_ & IS_SHUTDOWN)` and didn't notice the ` == 0;` part afterwards. Sorry for the false alarm :)